src/data/sum/order.lean

```diff
@@ -524,3 +527,61 @@ end,
   (sum_lex_dual_antidistrib α β).symm (inr (to_dual a)) = to_dual (inl a) := rfl
 
 end order_iso
+
+variable [has_le α]
+
+namespace with_bot
+
+/-- `with_bot α` is order-isomorphic to `punit ⊕ₗ α`, by sending `⊥` to `punit.star` and `↑a` to
+`a`. -/
+def order_iso_punit_sum_lex : with_bot α ≃o punit ⊕ₗ α :=
+⟨(equiv.option_equiv_sum_punit α).trans $ (equiv.sum_comm _ _).trans to_lex,
+begin
+  rintros (a | _) (b | _),
+  { simp },
+  { simp },
+  { simpa using with_bot.not_coe_le_bot a },
+  { simp }
+end⟩
+
+@[simp] theorem order_iso_punit_sum_lex_bot :
+  @order_iso_punit_sum_lex α _ ⊥ = sum.inl punit.star := rfl
```

Comment:
There's a `to_lex` missing on the RHS here. Same throughout (with `of_lex` as necessary).